src/main/asciidoc/_chapters/store_file_tracking.adoc

```diff
@@ -0,0 +1,175 @@
+:icons: font
+:experimental:
+
+== Overview
+
+This feature introduces an abstraction layer to track store files still used/needed by store
+engines, allowing for plugging different approaches of identifying store
+files required by the given store.
+
+Historically, HBase internals have relied on creating hfiles on temporary directories first, renaming
+those files to the actual store directory at operation commit time. That's a simple and convenient
+way to separate transient from already finalised files that are ready to serve client reads with data.
+This approach works well with strong consistent file systems, but with the popularity of less consistent
+file systems, mainly Object Store file systems, dependency on rename operations starts to introduce
+performance penalties. Amazon S3 Object Store, in particular, has been the most affected deployment,
+due to the its lack of atomic renames, requiring an additional locking layer implemented by HBOSS,
+to guarantee consistency and integrity of operations.
+
+With *Store File Tracking*, decision on where to originally create new hfiles and how to proceed upon
+commit is delegated to the specific Store File Tracking implementation.
+It can be set at individual Table or Column Family configurations, as well as in processes
+*hbase-site.xml* configuration file.
+
+NOTE: When specified in *hbase_site.xml*, this configuration is also saved into tables configuration
+at table creation time. This is to avoid dangerous configuration mismatches between processes, which
+could potentially lead to data loss.
+
+== Available Implementations
+
+Store File Tracking initial version provides three builtin implementations:
+
+* DEFAULT
+* FILE
+* MIGRATION
+
+### DEFAULT
+
+As per the name, this is the Store File Tracking implementation used by default when now explicit
+configuration has been defined. The DEFAULT tracker implements the standard approach using temporary
+directories and renames.
+
+### FILE
+
+A file tracker implementation that creates new files straight in the store directory, avoiding the
+need for rename operations. It keeps a list of committed hfiles in memory, backed by meta files, in
+each store directory. Whenever a new hfile is committed, the list of _tracked files_ in the given
+store is updated and a new meta file is written with this list contents, discarding the previous
+meta file now containing an out dated list.
+
+### MIGRATION
+
+A special implementation to be used when swapping between Store File Tracking implementations on
+pre-existing tables that already contain data, and therefore, files being tracked under an specific
+logic.
+
+== Usage
+
+For fresh deployments that don't yet contain any user data, *FILE* implementation can be just set as
+value for *hbase.store.file-tracker.impl* property in global *hbase-site.xml* configuration, prior
+to the first hbase start. Omitting this property sets the *DEFAULT* implementation.
+
+### Switching implementations globally
```

Comment:
I think Wellington means a migration case, rather than a table which already has an SFT implementation set.

Reply:
> Have you tried this operation? I do not think it works in this way...
> 
> The global config will only effect new tables. So I think here you need to alter the tables one by one...

Partially true. As Josh said, this is possible only when no explicit SFT configuration has yet been set in either hbase-site or the table descriptor. So basically a migration from the DEFAULT tracker to FILE. Let me update this section to make it clear this only works when these mentioned conditions are met.

Reply:
Oh, looking at the code, MigrateStoreFileTrackerProcedure will not always set the SFT impl to default, if we set a MIGRATION store file tracker globally, RollingUpgradeChore will set the SFT implementation to MIGRATION.

But this does not work actually, the trick here is that we will not actually reopen all the regions in MigrateStoreFileTrackerProcedure, because we think that we does not change anything actually. So this should be a bug, we should not set the SFT implementation to anything other than DEFAULT...

So for me, first, we could implement a special admin API to change the SFT implementation, to hide the intermediate MIGRATION state. This can be done with a special procedure which schedule two ModifyTableProcedure as sub procedures. And we could also implement a special migration tool, for migrating all the tables to specific SFT implementation after upgrading. I think this is a commom requirements for our users who deploy HBase on top of S3.

But anyway, I do not think the migration way described here can fully work, and it is very dangrous which may cause data loss(using a wrong SFT implementation will likely to cause data loss...)

Reply:
Ok, my test deployment didn't include HBASE-26263, which would basically cause the global config switch to be ignored. 

>MigrateStoreFileTrackerProcedure will not always set the SFT impl to default, if we set a MIGRATION store file tracker globally, RollingUpgradeChore will set the SFT implementation to MIGRATION.

We could add a validation in StoreFileTrackerFactory to not allow setting MIGRATION when no explicitly SFT config has been set yet? That would prevent this issue.

And yeah allowing MIGRATION to be set globally, does require for the MIGRATION SRC and DST be explicitly set to DEFAULT and FILE, respectively. Let me remove this section altogether, and update the next one to emphasise that switches are currently only supported at Table or CF level config.

A migration tool is highly desired, and it was on our plan to work on it once an initial version of this feature get released.

Reply:
Filed HBASE-26586, HBASE-26587 and HBASE-26588 for these things. Will work on them soon once the feature branch is merged :)